Update forms.mdx (#2665)
grommet/hpe-design-system: aries-site/src/pages/templates/forms.mdx

  <SimpleSignUpExample />
</Example>

> ### Jump to section...

Reviewer: Maybe we should omit the "Jump to section" on this PR since we don't have a design direction and leave that to be handled by separate intern project work.

Author: It might take a while for Aditi and I to develop a standardized look for the table of contents (i.e. Jump to section). In the meantime I would really want to see what the reaction is to at least the concept.

> [Use cases and example code](#when-to-use)  

Reviewer: Would all of the "sign-up", "sign-in" etc be considered use cases?

Author: Didn't quite understand this. Are they not real-world examples? Is that what use cases are?

> [Do's and Don'ts visual examples](#Dos)  

Reviewer 2: since we dont have visuals in this page should we change this name?

Author: We are working on this. I'm trying to get the text approved first, then we can work on the visuals.

[Use case and example code](use-cases)  
&nbsp;&nbsp;&nbsp;&nbsp;[Sign up](#sign-up)  
&nbsp;&nbsp;&nbsp;&nbsp;[Sign in](#sign-in)  
&nbsp;&nbsp;&nbsp;&nbsp;[Filter](#filter)  
&nbsp;&nbsp;&nbsp;&nbsp;[Sort](#sort)  
&nbsp;&nbsp;&nbsp;&nbsp;[Change password](#change-password)   
&nbsp;&nbsp;&nbsp;&nbsp;[Settings](#settings)  
&nbsp;&nbsp;&nbsp;&nbsp;[Pay](#pay)  
&nbsp;&nbsp;&nbsp;&nbsp;[Shipping](#shipping)   
&nbsp;&nbsp;&nbsp;&nbsp;[Customize](#customize)  
[Accessibility](#accessibility)   
[Exceptions and how to address them](#exceptions-and-how-to-address-them)  

> ### When to use

Reviewer: ```suggestion
## When to use
```

Reviewer 3: @dahliaferrer - FYI, all the heading/# comments @taysea has left is regarding applying the [semantically correct headings](https://design-system.hpe.design/foundation/typography?q=semantic#semantic-usage-of-heading-levels) to the content. Let us know if you have questions.

To efficiently capture information necessary for completing a task.

> ### When not to use

Reviewer: ```suggestion
## When not to use
```

Complicated processes where the user might need guidance at various stages, or when the form would be intimidatingly long. In these cases use a [Wizard](https://design-system.hpe.design/templates/wizard?q=error#guidance).

> #### Do's

Reviewer: ```suggestion
## Do's
```

> - Ask only for the required information. Keep the fields to a minimum to allow the user to finish their task as quickly as possible.

Reviewer: This is a very long list, is there any way we can break these into subcategories?

Reviewer 4: Example would be...text field, label text, error text, filled text field,

Author: @taysea yes @bezAdane  mentioned that, too. Will be doing that very soon :)

- Place the submit, reset and cancel buttons at the bottom left of the form. [See responsive button behavior](https://design-system.hpe.design/components/button?q=button#responsiveness). 
- Use specific labels on buttons that tell the user what action to expect 
- Provide feedback when the form is submitted successfully (e.g. notification, success page, green check mark, etc.) 
- [In the case of an error](#requiredoptional), red highlight the field where the error occurred 
- List below the button all relevant errors and the corrective action the user must take 
- Allow user to continue navigating through the application immediately

#### Don'ts

- Have more than two optional fields  
- Repeat the field label to use as placeholder text 
- Use asterisks for sign up and sign in forms 
- Red asterisks stresses the user and actually decrease success in filling out the form 
- Use generic labels on buttons, such as “Submit” 
- Restrict resource naming. Allow spaces, punctuation and unicode characters (ex. Björk Guðmundsdóttir, 联想, John Q. Public, etc.)
- No need to list requirements up front that the user is unlikely to commit (ex. double hyphens, starting with a space, etc.) 

> ### Single field forms

Reviewer: Is this an "exception"?

Author: I'm trying to make it easy for users to just go to one section where we list all of the guidance in ideal circumstances. The opposite of this is forcing the user to jump around to different sections to have to figure out what the "do's" are.

  <SimpleSignUpExample />
</Example>

### Jump to section...

> [When to use/not to use](#use)</br>

Reviewer 3: `<br />` is unnecessary. In Markdown, a new line should automatically occur with Return/Enter key. If that doesn't happen, add two spaces at the end of the line, then hit enter. The two spaces tells markdown to do the break row. Usually that is not necessary.

Update the jump to link to read `#when-to-use` (see corresponding comment below regarding auto creation of heading ids)

```suggestion
[When to use/not to use](#when-to-use)
```

Author: Yeah, it didn't seem to be accepting the line break, so I used the html code. Someone in the wwwworld suggested that double spaces aren't as good as the <br> anyway bc there are some people who still use double spaces after a full stop. But good call.

Reviewer 3: If you do use `br`, make sure the `/` comes after the `br`. I think it is reversed and causing parsing issues.

Reviewer 2: strange I attempted to fix this however still not doing the line break either my markdown skills are rusty or Im tired 😴  @halocline if you can take a stab I fixed the links so the preview shows. I tried `\` `two spaces` and `<br />` and still wasn't creating the line break 😕

Reviewer 3: I think this is a quirk unique to this particular markdown parser. 

Unfortunately, I don't think we can get the desired presentation with pure markdown. I will propose an alternative.

Reviewer 2: okay thank you Im glad it wasnt just me!

> [Do’s and Don’ts visual examples](#dos)</br>

Reviewer 3: These two comments: 1) rely on markdown generated ids, and 2) use either return or 2 spaces + return can be applied throughout.

Author: huh? 😂

## Do's

**Required fields**
- Ask only for the required information. Keep the fields to a minimum to allow the user to finish their task as quickly as possible.
- List only the most common field requirements below the relevant field. 
- Use asterisks for required fields when the form has more than two required fields and/or has optional fields. 
- Asterisks for required fields should be the same color as label field. 
- If a field is optional, write "Optional" next to the form field label.

**Format and fields**
- User-entered input must be encoded to prevent [SQL injection](https://www.w3schools.com/sql/sql_injection.asp) and [XSS attacks](https://en.wikipedia.org/wiki/Cross-site_scripting).
- Wrap all input elements within a FormField `<FormField label="Username"><TextInput /></FormField>`  
- Reveal fields progressively so the user is not overwhelmed.
- Write in sentence case (capitalize the first letter and all following letters should be lowercase).  
- Fill in default values for as many fields as possible. 

> - Use clear, concise title labels for each field. 

Reviewer: ```suggestion
- Use clear, concise labels for each form field. 
```

- Write helpful placeholder text inside the fields. 

> - If the user would likely have doubts, include helpful text below the field, such as [password requirements](#signup) or [additional resources](#customize) on a field's context.

Reviewer: ```suggestion
- If the user will likely have doubts about how to complete an input, include helpful text below the field, such as [password requirements](#signup) or [additional resources](#customize) on a field's context.
```

> - Place the submit, reset, and cancel buttons at the bottom left of the form. [See responsive button behavior](https://design-system.hpe.design/components/button?q=button#responsiveness). 

Reviewer: ```suggestion
- Place the submit, reset, and cancel buttons at the bottom left of the form. [See responsive button behavior](https://design-system.hpe.design/components/button#responsiveness). 
```
removing query from link

- Use specific labels on buttons that tell the user what action to expect. 

**Actions after the user submits information**
- Provide feedback when the form is submitted successfully (e.g. notification, success page, green check mark, etc.). 
- [In the case of an error](#requiredoptional), red highlight the field where the error occurred. 

> - List below the button all relevant errors and the corrective action the user must take. 

Reviewer: In our examples, I believe we are doing errors inline with the formfield when possible and above the submit button otherwise.

Author: - List all relevant errors and the corrective action the user must take. Place in-line when possible. When not possible, list above the submit button.

Reviewer: Yes, perfect. Small thing -- we've been typing as "inline".

- Allow the user to continue navigating through the application immediately.

## Don'ts

**Required fields**

> - Use asterisks for sign up and sign in forms. 

Reviewer: ```suggestion
- Use asterisks for sign up and sign in forms. This can be implemented using `required={{ indicator: false }}`.
```

> - Red asterisks stresses the user and actually decrease success in filling out the form. 

Reviewer: ```suggestion
- Use red asterisks as this stresses the user and actually decrease success in filling out the form. 
```

Author: Made slight changes. The HP brand writing style guide asks us to avoid using "as."

> - No need to list requirements up front that the user is unlikely to commit (ex. double hyphens, starting with a space, etc.). 

Reviewer: What does this mean?

Author: Edgar mentioned this. Do not list rejection criteria on the form if the user is unlikely to type them, such as double hyphens, starting with a space, etc. If they happen to violate the requirements by typing a double hyphen or starting with a space, THEN you give them an error and list the infraction. Do you think the user would understand that?

Author: - List rejection criteria that the user is unlikely to commit (ex. double hyphens, starting with a space, etc.). Only when they commit the error do you alert them.

**Format and fields**
- Have more than two optional fields. 

> - Repeat the field label to use as placeholder text. 

Reviewer: ```suggestion
- Repeat the field label as placeholder text. 
```

> - Use generic labels on buttons, such as “Submit.” 

Reviewer: What should be used instead?

Reviewer: "Submit" upon first impression falls into the category of "Save" -- but I can see how it would be considered unspecific. Maybe wording it as "Instead of generic labels such as "Submit", use "Save", "Publish"..."

  <SimpleSignUpExample />
</Example>

## When to use

To efficiently capture information necessary for completing a task.

## When not to use

> Complicated processes where the user might need guidance at various stages, or when the form would be intimidatingly long. In these cases use a [Wizard](https://design-system.hpe.design/templates/wizard?q=error#guidance).

Reviewer: A Wizard is still build using a form, so I'm wondering how we might clarify that.

Author: Think of it from the user's standpoint: we are giving them an alternative to using ONE basic form. A wizard.

## Do's

> **Required fields**

Reviewer: I think this (and the other sub categories should be h3

> **Required fields**

Author: Hm, is this still relevant or have we changed things already? I am just trying to make the subcategories of Don'ts smaller than the Don'ts title.

> **Required fields**

Reviewer: If we consider "required fields" a sub-category of "Do's", then I think it should be an h3

- Ask only for the required information. Keep the fields to a minimum to allow the user to finish their task as quickly as possible.
- List only the most common field requirements below the relevant field 
- Use asterisks for required fields when the form has more than two required fields and/or has optional fields. [See HPE UX research](https://www.figma.com/file/3fkwBelW5UsCbfwdDCJkT8/HPE-Form-Templates?node-id=4988%3A3653).  
- Asterisks for required fields should be the same color as label field 
- If a field is optional, write "Optional" in italics next to the form field label

**Format and fields**
- User-entered input must be encoded to prevent [SQL injection](https://www.w3schools.com/sql/sql_injection.asp) and [XSS attacks](https://en.wikipedia.org/wiki/Cross-site_scripting).
- Wrap all input elements within a FormField `<FormField label="Username"><TextInput /></FormField>`  

> - Reveal fields progressively so the user is not overwhelmed 

Reviewer: ```suggestion
- Reveal fields progressively so the user is not overwhelmed .
```
We have been following the convention of using a `.` at the end of bullet points. Should we update all bullets here to be consistent

- Write in sentence case (capitalize the first letter and all following letters should be lowercase)  
- Fill in default values for as many fields as possible 
- Use clear, concise title labels for each field 
- Write helpful placeholder text inside the fields 
- If the user would likely have doubts, include helpful text below the field, such as [password requirements](#signup) or [additional resources](#customize) on a field's context
- Place the submit, reset and cancel buttons at the bottom left of the form. [See responsive button behavior](https://design-system.hpe.design/components/button?q=button#responsiveness). 
- Use specific labels on buttons that tell the user what action to expect 

**Actions after the user submits information**
- Provide feedback when the form is submitted successfully (e.g. notification, success page, green check mark, etc.) 
- [In the case of an error](#requiredoptional), red highlight the field where the error occurred 
- List below the button all relevant errors and the corrective action the user must take 
- Allow the user to continue navigating through the application immediately

## Don'ts

**Required fields**

> - Have more than two optional fields 

Reviewer: This seems like we shouldn't have more than two optional fields in an initial task flow including a form (such as sign-up, etc.) -- but we also talk about how we can support more than 2 optional fields in something like profile settings (location, role, preferred contact method, phone number, secondary email, etc.). Should we clarify?

Reviewer: Also this bullet point doesn't seem to make sense under "Required Fields" section

Reviewer 4: Also not sure if want to limit users not to have options ( it is use case base ..

Author: I moved the "more than 2" to the "Fields" category below. And, again, the Do's and Don'ts are the ideal scenario. Any "exceptions," "edge cases," etc. can be addressed in the "Exceptions" section below.

- Use asterisks for sign up and sign in forms 

> - Red asterisks stresses the user and actually decrease success in filling out the form 

Reviewer: Should we format these with verbs. Something more like https://design-system.hpe.design/components/pageheader?q=page#subtitle -- wording like:

Required fields should not:
- Use asterisks for sign up and sign in forms
- Use red asterisks as this stresses the user ...

- No need to list requirements up front that the user is unlikely to commit (ex. double hyphens, starting with a space, etc.) 

**Format and fields**
- Repeat the field label to use as placeholder text 
- Use generic labels on buttons, such as “Submit” 

**Actions after the user submits information**

> - Restrict resource naming. Allow spaces, punctuation and unicode characters (ex. Björk Guðmundsdóttir, 联想, John Q. Public, etc.)

Reviewer: This bullet doesn't seem related to submission. It feels like it should fall under "Format"

Reviewer 4: This should be kept /fall under users-input guidance

Author: Ok, I can see your point. Note: I only put this in that category only because the user often doesn't discover that there was a problem with how they named their resource until after they've clicked on "submit," but I see that there's a case for putting this under the "requirements," category. Thx!

## Single field forms

- For [single form fields](#single-required-formfield) (ex: delete confirmation request), use an asterisk to denote that it is a "required" field. This can avoid user confusion and reduce the chance of "required" input error.

> <ExampleImagePreview

Reviewer: This example needs to be moved, it's not related to Single Field forms. Let me know if you think this quick link is helpful stilll and where you'd like to move it. I can move it for you.

Author: Has this been resolved? I looked for it and it seems we've worked it out of the document...?

Reviewer: This is still here. Can you let me know if you'd like it deleted or moved somewhere else on the page?

- Provide feedback when the form is submitted successfully (e.g. notification, success page, green check mark, etc.).
- [In the case of an error](#requiredoptional), red highlight the field where the error occurred.
- List all errors and the corrective action the user must take. Place inline when possible. Otherwise, list above the submit button.
- Allow the user to continue navigating through the application immediately.

## Don'ts

**Required fields**

- Use asterisks for sign up and sign in forms. This can be implemented using `required={{ indicator: false }}`.
- Use red asterisks, because has shown to actually decrease success in filling out the form.
- List rejection criteria that the user is unlikely to input (ex. double hyphens, starting with a space, etc.). Only alert them when they type the error.

**Format and fields**

- Have more than two optional fields.
- Repeat the field label as placeholder text.
- Use generic labels on buttons, such as “Yes," or "No." Instead use labels such as "Save," "Publish," "Create Project," etc.
- Restrict resource naming. Allow spaces, punctuation and unicode characters (ex. Björk Guðmundsdóttir, 联想, John Q. Public, etc.).

> ## Terms and Conditions

Author: In trying to come up with a simpler format, this seemed to fall more here than any other subtitle/category.

  alt="HPE Design System Form example where required and optional fields are used."
/>

> ### Use cases

Reviewer: ```suggestion
## Use cases
```

> ## Sign up

Reviewer: ```suggestion
### Sign up
```

  <SignUpExample />
</Example>

> ## Sign in

Reviewer: ```suggestion
### Sign in
```

  <SortExample />
</Example>

> ## Change password

Reviewer: ```suggestion
### Change password
```

  <ShippingExample />
</Example>

> ## Required vs optional fields

Reviewer: ```suggestion
### Required vs optional fields
```

  <RequiredFieldsExample />
</Example>

> ## Single required FormField

Reviewer: ```suggestion
### Single required FormField
```

>
  <CustomizeExample />
</Example>

## Accessibility

All Form input elements need to be contained within a FormField.

**FormFields must include:** 

- A "for" property that matches the id of its child input. If using Grommet, this is applied with the htmlFor prop.
- A "name" property applied on both the FormField and its child input element
- A clear, concise label

## Exceptions and how to address them

> - A login/registration form, which typically shows two to three required fields do not need to be annotated in any way - the form is short and the necessity of completing the fields is relatively obvious.

Reviewer: ```suggestion
- A login/registration form, which typically shows two to three required fields do not need to be annotated in any way - the form is short and the necessity of completing the fields is standard practice.
```

- For login/registration forms, apply `required={{ indicator: false }}` to the FormFields to remove the required indicator.

- Forms with minimal fields are optimal but occasionally, capturing extra information would be highly beneficial for the user. The following are some options for addressing these cases: 

> - Create the form the optimal way: with only the minimum required fields so that the user can focus on completing their task. Separately, in a preferences or settings dashboard, for example, create those fields where the user can proactively add this helpful information at another time. 

Reviewer: ```suggestion
   - Create the form the optimal way: with only the minimum required fields so that the user can focus on completing their task. Separately, in a preferences or settings dashboard, for example, create those fields where the user can proactively add this helpful information at another time. 
```
This feels like it should be a sub-bullet of the bullet above

> - However, add the optional field if there is compelling evidence that the user’s experience would be greatly improved if they could input this extra information in this initial form. Ensure to mark the field as optional. 

Reviewer: Can we clarify an example of when this would be? I'm having a hard time thinking of one.

Author: @taysea  Just throwing it out there: how about asking for their title or role on a sign-up page? As we know, even something as benign as asking for a title or role can be sticky. A user might hesitate for a myriad of reasons. It's better to not ask them for this info up front, but allow them to add it later in their profile settings.

Reviewer: Got it, I think that's a helpful example to add here. Something like: "For example, it may be unnecessary to ask a user for their role on a sign-up page, but including it as an optional field in profile settings is reasonable."

Author: Check if it makes sense now :)

> <ExampleImagePreview

Reviewer: This example should be moved to the section with related guidance

Author: Is this where you expect it to be now? I've moved everything into categories, so I hope that it fits the category where you'd expect to find this.

  <SingleFormFieldExample />
</Example>

### Customize

<Example
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/forms/CustomizeExample.js"
  figma="https://www.figma.com/file/3fkwBelW5UsCbfwdDCJkT8/HPE-Form-Templates?node-id=54%3A608"
>
  <CustomizeExample />
</Example>

### Accessibility

All form input elements must be contained within a FormField.

**FormFields must include:** 

> - A "for" property that matches the id of its child input. If using Grommet, this is applied with the htmlFor prop.

Reviewer: ```suggestion
- A "for" property that matches the `id` of its child input. If using Grommet, this is applied with the `htmlFor` prop.
```

- A `name` property applied on both the FormField and its child input element.
- A clear, concise label.

## Exceptions and how to address them

- A login/registration form, which typically shows two to three required fields do not need to be annotated in any way - the form is short and the necessity of completing the fields is standard practice.

   - For login/registration forms, apply `required={{ indicator: false }}` to the FormFields to remove the required indicator.

> - Forms with minimal fields are optimal but occasionally, capturing extra information would be highly beneficial for the user. The following are some options for addressing these cases: 

Reviewer: ```suggestion
- Forms with minimal fields are optimal, but occasionally, capturing extra information would be highly beneficial for the user. The following are some options for addressing these cases: 
```

>    - Create the form the optimal way; with only the minimum required fields so that the user can focus on completing their task. Separately, in a preferences or settings dashboard, for example, create those fields where the user can proactively add this helpful information at another time. An example of a helpful but not required field could be asking for someone's official title or role. It might be highly useful to capture it right away on a sign-in form, but a title or role might be too problematic for the admin to enter. Instead, offer this "role" or "title" field in profile settings.

Reviewer: ```suggestion
   - Create the form the optimal way; with only the minimum required fields so that the user can focus on completing their task. Separately, in a preferences or settings dashboard, for example, provide a form that allows the user to add helpful, optional information at another time. An example of a helpful but not required field could be asking for someone's official title or role. If it is not necessary to capture this information right away on a sign-in form, offer this "role" or "title" field in profile settings.
```

  <SingleFormFieldExample />
</Example>

### Customize

<Example
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/forms/CustomizeExample.js"
  figma="https://www.figma.com/file/3fkwBelW5UsCbfwdDCJkT8/HPE-Form-Templates?node-id=54%3A608"
>
  <CustomizeExample />
</Example>

### Accessibility

All form input elements must be contained within a FormField.

**FormFields must include:** 

> - A "for" property that matches the id of its child input. If using Grommet, this is applied with the htmlFor prop.

Reviewer: ```suggestion
- An `for` property that matches the `id` of its child input. If using Grommet, this is applied with the `htmlFor` prop.
```

> - A "name" property applied on both the FormField and its child input element

Reviewer: ```suggestion
- A `name` property applied on both the FormField and its child input element
```

- A clear, concise label

## Exceptions and how to address them

- A login/registration form, which typically shows two to three required fields do not need to be annotated in any way - the form is short and the necessity of completing the fields is standard practice.

> - For login/registration forms, apply `required={{ indicator: false }}` to the FormFields to remove the required indicator.

Reviewer: This should be a sub-bulllet of the above.

Author: I think this has changed since you wrote this.

  <SingleFormFieldExample />
</Example>

### Customize

<Example
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/forms/CustomizeExample.js"
  figma="https://www.figma.com/file/3fkwBelW5UsCbfwdDCJkT8/HPE-Form-Templates?node-id=54%3A608"
>
  <CustomizeExample />
</Example>

### Character counter

**When to use**  
In cases where the user is given an opportunity to express themselves or would likely exceed a field's character limit. Examples include some description fields, feedback, feature requests, reporting issues, etc.

**Guidance**

- Place the character counter at the bottom right of the field.

> - [See guidance on character limits for fields](https://design-system.hpe.design/components/tag#anatomy). Note that your backend may have character limitations. When you can choose the character limits, however, set a generous number to avoid frustrating the user. A suggestion to ensure your limit is user-friendly, check the analytics to see how long it is taking your users to complete this task. Longer than usual time could indicate the character limits are too low. Increase accordingly.

Reviewer 3: - Internal links should not include the `https://design-system.hpe.design`.
- Hyperlinked portion should typically be focused/limited to the subject of the sentence.
- Link does not seem like it is pointing to the correct content.

```suggestion
- See guidance on [character limits for fields](/components/tag#anatomy). Note that your backend may have character limitations. When you can choose the character limits, however, set a generous number to avoid frustrating the user. A suggestion to ensure your limit is user-friendly, check the analytics to see how long it is taking your users to complete this task. Longer than usual time could indicate the character limits are too low. Increase accordingly.
```

- Use a numerator/denominator format. Ex. 0/128.
- Show the progression in the numerator as the user is typing. Ex. 52/128 if the user has typed 52 characters, 53/128 when they type an extra character, 54/128, 55/128, etc.
- Use default text styling.
- Make the field roughly the same size as the character limit.
- Red highlight the field when the user meets the character limitation.

**Accessibility**

- Announce the max character allowed before the user begins typing.
- Announce when the user has typed 50% of the max allowed.

### Accessibility

> All form input elements must be contained within a FormField.

Reviewer 3: Why?

We have an opportunity to be educating here, and should be. This and the following "musts" reads as directives, but is lacking explanation for the importance of why this should be done.

Opportunity to explain why doing, what needs to be done, and what it accomplishes.

Reviewer 3: Something like this might work:

"Forms need to be presented in an accessible manner for all users. Accordingly, all form inputs need to be accompanied by a label which identifies itself to screen readers. Placing all inputs within a Grommet FormField and applying the following play a big role in ensuring the inputs are accessible:"

**FormFields must include:**

> 

Reviewer 3: It is also probably worth calling out "placeholder text is not an accessible substitute for an inputs label"

Author: Added it!

- A "for" property that matches the `id` of its child input. If using Grommet, this is applied with the `htmlFor` prop.
- A `name` property applied on both the FormField and its child input element.

> - A clear, concise label.

Reviewer 3: This is probably some of the most, if not the most, important content on the entire page. It is one of the primary accessibility offenders.

Consider a more prominent placement.

Author: I've moved the accessibility section below the Don'ts. From what I remember, most of our guidance pages have it near the bottom. But putting these considerations closer to the do's and don'ts might help emphasize that accessibility is a positive for our designs and everyone's experience in general.

> - A clear, concise label.

Reviewer 3: A code example accompanying this guidance could be useful.

Reviewer: @dahliaferrer We can add:

```
<FormField label="Username" htmlFor="example-username" name="example-username"?
   <TextInput id="example-username" name="example-username" />
</FormField>
```

Author: Ok, I will add it, but question: is this necessary if a developer is going to create an interactive example?

Reviewer: You can just add the code snippet for now using ``` markdown syntax before and after the code block. I think in this case we want to be focused on the code itself, so showing it inline is a good approach.

  <CustomizeExample />
</Example>

### Character counter

**When to use**  
In cases where the user is given an opportunity to express themselves or would likely exceed a field's character limit. Examples include some description fields, feedback, feature requests, reporting issues, etc.  

**Guidance**
- Place the character counter at the bottom right of the field.  

> - [See guidance on character limits for fields](https://design-system.hpe.design/components/tag#anatomy). If you have doubts about how many characters to limit in the field, set a generous limit so as to not frustrate the user. When possible, check analytics to see how long it is taking your users to complete this task. Longer than usual time could indicate the character limits are too few. Increase accordingly.

Reviewer: Sometimes the character count is limited by the backend -- wondering if we should mention that

Author: Better? I figure if the backend has limitations, maybe it's for security or data volume purposes and we shouldn't encourage them to mess with that LOL but you tell me. 

"Note that your backend may have character limitations. When you can choose the character limits, however, make it generous to avoid frustrating the user."

Reviewer: Perfect.

- Use a numerator/denominator format. Ex. 0/128.  

> - Count down the characters as the user is typing. Ex. 52/128 if the user has typed 52 characters. 53/128 when they type an extra character, etc.  

Reviewer: This example feels like counting up to be

> - The character counter should be the same color as the field title.  

Reviewer: We can just say "text" as the color

Author: Could you write the whole sentence? I didn't quite catch what you meant :)

Reviewer: The character counter should be `text` color.

Author: - The character counter should use the `text` styling (same color, weight, font, etc.)

Reviewer: `text` was referring solely to the color. If we are saying text styling in general, we can just say:

- The character counter should use default text styling.

- Make the field roughly the same size as the character limit.  

> - Red highlight the field when the user meets the character limitation.  

Reviewer: Typically we call this "Error validation" instead of red highlight.

Author: Error validation could look very differently in different design systems. I figure "red highlight the field" is specific and is in line with our design system.

Reviewer: Maybe specify with something like "When the user meets the character limit, use the form field error validation, which appears as a red highlight."

We define a lot of the color/states by default and try to inform people in guidance of these default treatments, so my suggestion is to try to bring the connection that the red highlight is the HPE Design System default

Reviewer 5: Yep, we try to avoid people coming up with their own "red" and instead explaining it more semantically and tied to the properties and components involved.

**Accessibility**
- Announce the max character allowed before the user begins typing.
- Announce when the user has typed 50% of the max allowed. 

<Example

> </Example>

Reviewer: We can remove the example tag for now. This will be added by a dev later on.
